src/nvim/api/vimscript.c

```diff
@@ -745,3 +746,294 @@ Dictionary nvim_parse_expression(String expr, String flags, Boolean highlight, E
+{
+  Dictionary dict = ARRAY_DICT_INIT;
+  kv_resize(dict, details ? 7 : 2);
+
+  PUT(dict, "name", STRING_OBJ(name));
+  PUT(dict, "type", CSTR_TO_OBJ("builtin"));
+
+  if (details) {
+    PUT(dict, "min_argc", INTEGER_OBJ(fn->min_argc));
+    PUT(dict, "max_argc", INTEGER_OBJ(fn->max_argc));
+    PUT(dict, "base_arg", INTEGER_OBJ(fn->base_arg));
+    PUT(dict, "fast", BOOLEAN_OBJ(fn->fast));
+
+    // Argument names
+    if (fn->argnames != NULL) {
+      Array overloads = ARRAY_DICT_INIT;
+      Array names = ARRAY_DICT_INIT;
+      char buf[64] = { 0 };
+      size_t pos = 0;
+
+      // Argument are separated with unit separator (ascii \x1F),
+      // overloads with record separator (ascii \x1E).
+      for (const char *p = fn->argnames;; p++) {
+        if (*p == NUL || *p == '\x1F' || *p == '\x1E') {
+          assert(pos < 64);
+          buf[pos++] = NUL;
+          String str = {
+            .data = xmemdupz(buf, pos),
+            .size = pos - 1,
+          };
+          pos = 0;
+
+          if (*p == '\x1F') {  // unit separator: argument separator
+            kv_push(names, STRING_OBJ(str));
+          } else if (*p == '\x1E') {  // record separator: overload separator
+            kv_push(names, STRING_OBJ(str));
+            kv_push(overloads, ARRAY_OBJ(names));
+            names = (Array)ARRAY_DICT_INIT;
+          } else {  // NUL
+            kv_push(names, STRING_OBJ(str));
+            kv_push(overloads, ARRAY_OBJ(names));
+            break;
+          }
+        } else {
+          buf[pos++] = *p;
+        }
+      }
+
+      PUT(dict, "argnames", ARRAY_OBJ(overloads));
+    }
+  }
+
+  return dict;
+}
+
+/// Lists Vimscript functions.
+///
+/// @param query    When string gets information about the function under this name.
+///                 When dictionary lists functions. The following keys are accepted:
+///                 - builtin (boolean, default false) Include builtin functions.
+///                 - user    (boolean, default true) Include user functions.
+/// @param opts     Options dictionary:
+///                 - details (boolean) Include function details.
+///                 - lines   (boolean) Include user function lines. Ignored for builtin
+///                           functions.
+/// @param[out] err Error details, if any
+/// @return A dictionary describing a function when {query} is a string, or a map of
+///         function names to dictionaries describing them when {query} is a dictionary.
+Dictionary nvim_get_functions(Object query, Dictionary opts, Error *err)
+  FUNC_API_SINCE(10)
+{
+  Dictionary rv = ARRAY_DICT_INIT;
+
+  if (query.type != kObjectTypeString && query.type != kObjectTypeDictionary
+      && (query.type == kObjectTypeArray && query.data.array.size != 0)) {
+    api_set_error(err, kErrorTypeValidation, "query is not a dictionary or string %d", query.type);
+    return rv;
+  }
+
+  bool details = false;
+  bool lines = false;
+  for (size_t i = 0; i < opts.size; i++) {
```

Comment:
can avoid this for-loop by updating `keysets.lua`
https://github.com/neovim/neovim/blob/6cc6e11929ad76a2dc5204aed95cb9ed1dafde23/src/nvim/api/extmark.c#L618